src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -8,39 +10,58 @@ module ZkFold.Symbolic.Data.UInt (
 import           Data.List                                                 (map, unfoldr, zip, zipWith)
 import           Data.Map                                                  (fromList, (!))
 import           Data.Traversable                                          (for, traverse)
 import           Data.Tuple                                                (swap)
-import           GHC.Natural                                               (Natural, naturalFromInteger)
-import           Prelude                                                   (Integer, error, flip, otherwise, return,
-                                                                            ($), (++), (.), (>>=))
+import           GHC.Generics                                              (Generic)
+import           GHC.Natural                                               (naturalFromInteger)
+import           GHC.TypeNats                                              (Natural)
+import           Prelude                                                   (Integer, concatMap, error, flip, foldl, fst,
+                                                                            otherwise, return, snd, ($), (++), (.),
+                                                                            (<>), (>>=))
 import qualified Prelude                                                   as Haskell
 import           Test.QuickCheck                                           (Arbitrary (..), chooseInteger)
 
 import           ZkFold.Base.Algebra.Basic.Class
-import           ZkFold.Base.Algebra.Basic.Field                           (Zp, fromZp)
+import           ZkFold.Base.Algebra.Basic.Field                           (Zp, fromZp, toZp)
 import           ZkFold.Base.Algebra.Basic.Number
-import           ZkFold.Prelude                                            (length, replicate, replicateA, splitAt,
-                                                                            take)
+import           ZkFold.Prelude                                            (drop, length, replicate, replicateA,
+                                                                            splitAt, take, (!!))
 import           ZkFold.Symbolic.Compiler                                  hiding (forceZero)
 import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Combinators    (expansion, splitExpansion)
 import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.MonadBlueprint
 import           ZkFold.Symbolic.Data.Bool
 import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.Data.Conditional
 import           ZkFold.Symbolic.Data.Eq
 import           ZkFold.Symbolic.Data.Eq.Structural
+import           ZkFold.Symbolic.Data.Ord
 
 -- TODO (Issue #18): hide this constructor
 data UInt (n :: Natural) a = UInt ![a] !a
-    deriving (Haskell.Show, Haskell.Eq)
+    deriving (Haskell.Show, Haskell.Eq, Generic, NFData)
+
+class DivMod a where
```

Comment:
This is a Euclidean domain.

Reply:
Perhaps should be in the ZkFold.Basic...Algebra hierarchy like

```Haskell
class Semiring a => EuclideanDomain a where
```

Then we can have an extended Euclidean algorithm for all Euclidean domains like `Natural`, `Integer`, `Zp` and `UInt n a`.